// node-discordbots/library/base/DiscordBotsList.js
const snekfetch = require("snekfetch");

module.exports = class DiscordBotsList {
	
	constructor(options) {
		this.token = options.token;
		this.client = options.clientID;
	}

	grabUser(user) {
		return this._get(`https://discordbots.org/api/users/${user}`);
	}

	grabBot(bot) {
		return this._get(`https://discordbots.org/api/bots/${bot}`);
	}

	grabVotes() {
		return this._get(`https://discordbots.org/api/bots/${this.client}/votes`)
	}

	grabStats(bot) {
		return this._get(`https://discordbots.org/api/bots/${bot}/stats`)
	}

	postStats(guildCount) {
		return this._post(guildCount, `https://discordbots.org/api/bots/${this.client}/stats`)
	}

	_post(guildCount, base) {
    	return new Promise((resolve, reject) => {
        	snekfetch.get(base)
        		.set("Authorization", this.token)
        		.set("User-Agent", "DBots Library ( https://npmjs.org/package/dbots ) - Created by Doh")
        		.send({server_count : parseInt(guildCount)})
        		.then(res => {
            		if (res.status !== 200) return reject(res);
            		else return resolve(`Successfully posted ${guildCount} guilds to Discord Bots List.`);
        		}).catch((err) => {
            		reject(err);
       			});

		});	
	}

	_get(base) {
    	return new Promise((resolve, reject) => {
        	snekfetch.get(base)
        		.set("Authorization", this.token)
        		.set("User-Agent", "DBots Library ( https://npmjs.org/package/dbots ) - Created by Doh")
        		.then(res => {
            		if (res.status !== 200) return reject(res);
            		else return resolve(res.body);
        		}).catch((err) => {
            		reject(err);
       			});

		});	
	}
}
